Adds tests to verify error notices when adding out-of-stock products

BurakParsAydin opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #57318  .

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Run the following command:
`pnpm --filter=@woocommerce/block-library test:e2e tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts --debug`


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->
Add E2E test to verify error notice displays correctly when adding out-of-stock products from Product Collection
</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Tests**
  - Added an end-to-end test to ensure users see an error notice when trying to add out-of-stock products from a Product Collection in WooCommerce.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

BurakParsAydin committed:
Adds tests to verify error notices when adding out-of-stock products

BurakParsAydin committed:
Refactor the code

BurakParsAydin committed:
Add the changelog file

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

BurakParsAydin commented:
I'm getting the following error when running the command. Could I get some help with this?: 

![Screenshot from 2025-05-30 22-02-11](https://github.com/user-attachments/assets/be82ec11-2769-44d8-ba86-5d745ba97f89)

Aljullu reviewed (commented):
Thanks for working on this, @BurakParsAydin! Some things that might help:

1. The test file name should end in `.block_theme.spec.ts` (notice the `block_theme` part), this tells the e2e test tooling to run the test in a block theme. It's also possible to use `classic_theme` or `block_theme_with_templates` when needed, but in generic cases like this one, `block_theme` is the correct one to use.
2. I would avoid creating any new category during the tests, as this can slow down the test run significatively.
3. Similarly, I would avoid creating a new product and instead would try to use one of the existing ones. The state is reset between tests, so you can do any edits you need to them without worrying to affect follow-up tests.
4. As a tip, you can add `--debug` to the command you use to run tests and that will open a browser window. That's very helpful because you can see the test live and you can also open new tabs and, with that, navigate the site, see the categories, products, etc.

Hope that helps! :slightly_smiling_face:

BurakParsAydin committed:
Remove redundant changelog file

BurakParsAydin committed:
Update the file name

BurakParsAydin committed:
Merge branch 'trunk' into add-57318

BurakParsAydin committed:
Simplified test flow

Aljullu reviewed (commented):
Thanks for updating this, @BurakParsAydin!

> For the test, I only got stuck on how to capture the error when adding the product. Since it’s an out-of-stock product, there’s no "Add to Cart" button, so I can’t add it to the cart. Could you help me with this part?

You're right. I initially thought testing with an _Out of stock_ product was the easiest approach. That changes the _Add to Cart_ button with a _Read more_ link as you say. So instead of using an _Out of stock_ product, it might be better to test with a product sold individually. After pressing the _Add to Cart_ button twice, an error should be displayed.

BurakParsAydin committed:
Refactor and improve test coverage

BurakParsAydin commented:
@Aljullu Thank you for your help. It really helped me implement the test.
Good news: I've mostly completed the test and pushed the commit.

The only issue I’m facing is with the final step, which refers to [this](https://github.com/woocommerce/woocommerce/blob/3a7ef3359026e270aebbfbedaeb4c09d3ba97f89/plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts#L80-L83).

The script can't find `.wc-block-components-notice-banner.is-error` because the banner only appears inside the sidebar that opens when clicking the cart (bag) icon. If you get a chance to run the test, you’ll see what I mean.

Do you have any suggestions on how I could approach this part?

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,85 @@
+	} ) => {
+		await blockUtils.createManagedStockProduct();
+		await admin.createNewPost();
+		await editor.insertBlock( { name: 'woocommerce/single-product' } );
```
We should be testing the _Product Collection_ block (`woocommerce/product-collection`) instead of the _Single Product_ block, here.

Aljullu reviewed (commented):
I think the issue is that you were adding the wrong block to the page. You should test the _Product Collection_ block, not the _Single Product_ block. :slightly_smiling_face:  The latter doesn't have error notices itself, so it's expected they won't show up.

BurakParsAydin committed:
Add the store notices block

BurakParsAydin commented:
@Aljullu Thanks for your review 🙂 I've added the Store Notice block - please refer to my [latest commit](https://github.com/woocommerce/woocommerce/pull/58402/commits/ad0bc378cb0a65efba499cb0dfb861e3c4def9b4), which addresses the issue. 
The test is now running well on my end 🚀 Looking forward to your feedback

Aljullu commented:
Thanks for the updates, @BurakParsAydin! Not sure if you saw my [last comment](https://github.com/woocommerce/woocommerce/pull/58402#pullrequestreview-2909423078), but the issue is that we need to test the _Product Collection_ block (`woocommerce/product-collection`), not the _Single Product_ block (`woocommerce/single-product`). Your test [is adding the _Single Product_ block to the page](https://github.com/woocommerce/woocommerce/pull/58402/files#diff-205ca2b430ba5da692d90de898dd88731849a890f322ddd936dd3f85e615cafdR67) and testing it, instead of testing the _Product Collection_ block.

https://github.com/woocommerce/woocommerce/issues/57318 is about adding the test for the _Product Collection_ block, so this PR needs to be updated so the block that is added in the test is the _Product Collection_. Hope that makes sense!

BurakParsAydin committed:
Refactor the test

BurakParsAydin commented:
@Aljullu That was a helpful insight. I appreciate it. FYI, I adjusted the code. 
Waiting for your feedback.

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,81 @@
+		await expect( productCart ).toBeVisible();
+
+		// Add to cart once — succeeds.
+		await productCart.locator( '.add_to_cart_button' ).click();
```
It's usually a good practice to use [`.getByText()`, `.getByRole()`, etc. locators](https://playwright.dev/docs/locators) instead of using class names. In this case, could we use `.getByText( 'Add to cart' )`?

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,81 @@
+
+		// Verify error notice is displayed.
+		await expect(
+			page.locator( '.wc-block-components-notice-banner.is-error' )
```
Similarly, here it would be better to check for the error text message (or part of it) instead of the element class name.

BurakParsAydin replied:
@Aljullu I'm thinking to target this via `getByRole()`. What paremeter should I use? `dialog` doesn't work

Aljullu replied:
The role is `alert`, if I'm not wrong. :slightly_smiling_face: 

In this case, it might also be good to check that it contains an error message, not only that it's visible. Something like this, maybe?

```JS
await expect( page.getByRole( 'alert' ) ).toBeVisible();
await expect( page.getByRole( 'alert' ) ).toHaveText(
	/maximum quantity/i
);
```

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,81 @@
+
+		const productName = 'A Managed Stock';
+
+		await editor.publishAndVisitPost();
```
I wonder if we could replace all of this with the `pageObject` utils:

```JS
			await pageObject.createNewPostAndInsertBlock();
			await pageObject.publishAndGoToFrontend();
```

Like it's done in [this test](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts#L30), for example.

BurakParsAydin replied:
@Aljullu FYI, I've adjusted the test as per your request. There's just **this part** left I keep getting errors here. If possible, could you make a commit suggestion for this part?

Aljullu replied:
I left the comments inline. Hope that helps. :slightly_smiling_face:

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,81 @@
+		await wpCLI(
+			'wc product create --name="A Managed Stock" --regular_price=10 --manage_stock=true --stock_quantity=1 --user=admin'
+		);
+	}
```
Given that this method is only used in this test, I wonder if we should simply have it inline in the test. That would help simplify the code.

Aljullu reviewed (commented):
Thanks for the updates, @BurakParsAydin! The test is passing now and it's correctly testing that the error notices are visible in the _Product Collection_ block. :raised_hands: 

I left a few inline comments with some suggestions which I think would help make the test simpler and more reliable.

BurakParsAydin committed:
Refactor the test - 1st phase

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,50 @@
+		await expect( productCart ).toBeVisible();
+
+		// Add to cart once — succeeds.
+		await productCart.getByRole( 'button' ).click();
```
I think it will be more reliable if we make sure the button has the `Add to Cart` text:

```suggestion
		await productCart.getByRole( 'button', { name: 'Add to cart' } ).click();
```

BurakParsAydin committed:
Update plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts

Co-authored-by: Albert Juhé Lluveras <aljullu@gmail.com>

BurakParsAydin committed:
Refactor the test - 2nd phase

BurakParsAydin committed:
Merge branch 'trunk' into add-57318

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,53 @@
+	} ) => {
+		const productName = 'A Managed Stock';
+
+		wpCLI(
```
Heads-up that `wpCLI()` needs to be awaited:

```suggestion
		await wpCLI(
```

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,53 @@
+
+		await singleProduct.getByText( 'create your own' ).click();
+
+		await editor.publishAndVisitPost();
```
You can replace this with:

```suggestion
		await pageObject.createNewPostAndInsertBlock();
		await pageObject.publishAndGoToFrontend();
```

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,53 @@
+		admin,
+		editor,
+		page,
+	} ) => {
```
Then you must add the `pageObject` here:

```suggestion
const test = base.extend< { pageObject: ProductCollectionPage } >( {
	pageObject: async ( { page, admin, editor }, use ) => {
		const pageObject = new ProductCollectionPage( {
			page,
			admin,
			editor,
		} );
		await use( pageObject );
	},
} );

test.describe( 'Product Page: error notices when adding out-of-stock products', () => {
	test( 'displays error notice when attempting to add product beyond stock limit', async ( {
		page,
		pageObject,
	} ) => {
```

You will also need to add the import at the top of the file:

```TS

/**
 * Internal dependencies
 */
import ProductCollectionPage from './product-collection.page';
```

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,53 @@
+			.click();
+
+		// Add to cart again — triggers out-of-stock error.
+		await productCart.getByRole( 'button' ).click();
```
I would update this line that the one above, making sure the `button` has the `Add to cart` text.

BurakParsAydin replied:
@Aljullu Since the button text changes after the first click and says "1 in cart," I thought it might be better to leave it as is. Do you have any recommendations at this point?

Aljullu replied:
Ah, my bad, you're right. In that case, could we add `{ name: '1 in cart' }`?

What I don't like about simply using `getByRole( 'button' )` is that if any other button is added to the Product Collection block in the future, the test might fail because it could try to click on the wrong button. For example, imagine in the future an amplifier button is added to the product image, this test might fail because it won't know which button it needs to click. So, in general, it's better to be as specific as possible with this kind of selectors. Do you agree?

BurakParsAydin replied:
@Aljullu Thanks for the recommendation. I’ve updated that line, but I ended up targeting the button a bit differently. Hope that’s okay with you.

It seems` { name: '1 in cart' } `doesn’t work, since the button’s accessible name is actually "Add to cart."

Aljullu reviewed (commented):
Thanks for the updates, @BurakParsAydin, I left a few more comments, but it's shaping up good. :100:

BurakParsAydin committed:
Update plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts

Co-authored-by: Albert Juhé Lluveras <aljullu@gmail.com>

BurakParsAydin committed:
Update plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts

Co-authored-by: Albert Juhé Lluveras <aljullu@gmail.com>

BurakParsAydin committed:
Update plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts

Co-authored-by: Albert Juhé Lluveras <aljullu@gmail.com>

BurakParsAydin committed:
Tidying up

BurakParsAydin commented:
@Aljullu FYI, I’ve adjusted the test as you suggested. The only part left is the one related to the add-to-cart event you mentioned above. Looking forward to your feedback 🙂

BurakParsAydin committed:
Target the button with its text

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,59 @@
+		// Add to cart once — succeeds.
+		await productCart
+			.getByRole( 'button', { name: 'Add to cart' } )
+			.click();
```
@BurakParsAydin I think this is a good suggestion. Do you think we can add it to the test?

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,59 @@
+	},
+} );
+
+test.describe( 'Product Page: error notices when adding out-of-stock products', () => {
```
I would simplify this `describe`, in the test itself we already describe what the test is doing:

```suggestion
test.describe( 'Product Page: error notices', () => {
```

Aljullu reviewed (commented):
Thanks for the updates, @BurakParsAydin! I left a couple of last comments, but overall everything is looking good.

BurakParsAydin committed:
Update plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts

Co-authored-by: Albert Juhé Lluveras <aljullu@gmail.com>

BurakParsAydin committed:
Refactor the click event

BurakParsAydin committed:
Run npx prettier

BurakParsAydin commented:
@Aljullu Thanks! FYI, I've adjusted the test as per your request. Waiting for your feedback.

Aljullu commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-collection/product-collection-errors.block_theme.spec.ts:
```diff
@@ -0,0 +1,57 @@
+ * External dependencies
+ */
+import { Page } from '@playwright/test';
+import { test as base, expect, Editor, wpCLI } from '@woocommerce/e2e-utils';
```
`Page` and `Editor` aren't used in this test, so they shouldn't be imported. :slightly_smiling_face:

Aljullu reviewed (commented):
Thanks, @BurakParsAydin! Everything looks good to me. Apparently, there is only one issue remaining related to linting. I commented it inline. :point_down:

BurakParsAydin committed:
Tidy up the test: improve naming, remove await and remove unused import

BurakParsAydin commented:
@Aljullu Thanks for the review  🙂 I've adjusted it as you addressed. 
I also renamed the constant name for clarity, removed await - it doesn’t need to be awaited if `use` itself is synchronous internally I think.

If there’s a related task you think I can help with, feel free to let me know - I'd be happy to jump in!

Apart from that, just wanted to say a big thank you for all the guidance and support you gave me on this task 🙏🏼
Your detailed feedback and clear direction really helped me understand things better and move forward with confidence. I truly appreciate the time and effort you put into reviewing and mentoring - means a lot! ❤️

Aljullu reviewed (commented):
> removed await - it doesn’t need to be awaited if `use` itself is synchronous internally I think.

Apparently lint is complaining about that, so I think we need to add that `await` again. :smile: 

> Apart from that, just wanted to say a big thank you for all the guidance and support you gave me on this task 🙏🏼
> Your detailed feedback and clear direction really helped me understand things better and move forward with confidence. I truly appreciate the time and effort you put into reviewing and mentoring - means a lot! ❤️

Not at all, thanks a lot for your help and contribution to WooCommerce! :raised_hands:

BurakParsAydin committed:
Update product-collection-errors.block_theme.spec.ts

Aljullu reviewed (approved):
Awesome, thanks for all your work, @BurakParsAydin! This is looking good now, merging it into `trunk`! :partying_face:
